In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import re # regular expression
import nltk

import os

In [2]:
data=pd.read_csv("blogtext.csv")

In [3]:
data.head()

,id,gender,age,topic,sign,date,text
0,2059027,male,15,Student,Leo,"14,May,2004","Info has been found (+/- 100 pages,..."
1,2059027,male,15,Student,Leo,"13,May,2004",These are the team members: Drewe...
2,2059027,male,15,Student,Leo,"12,May,2004",In het kader van kernfusie op aarde...
3,2059027,male,15,Student,Leo,"12,May,2004",testing!!! testing!!!
4,3581210,male,33,InvestmentBanking,Aquarius,"11,June,2004",Thanks to Yahoo!'s Toolbar I can ...


In [4]:
data.shape

(681284, 7)

In [5]:
data.size

4768988

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 681284 entries, 0 to 681283
Data columns (total 7 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   id      681284 non-null  int64 
 1   gender  681284 non-null  object
 2   age     681284 non-null  int64 
 3   topic   681284 non-null  object
 4   sign    681284 non-null  object
 5   date    681284 non-null  object
 6   text    681284 non-null  object
dtypes: int64(2), object(5)
memory usage: 36.4+ MB


In [7]:
data.describe()

,id,age
count,6.812840e+05,681284.000000
mean,2.397802e+06,23.932326
std,1.247723e+06,7.786009
min,5.114000e+03,13.000000
25%,1.239610e+06,17.000000
50%,2.607577e+06,24.000000
75%,3.525660e+06,26.000000
max,4.337650e+06,48.000000


In [8]:
data.isna().any()

id        False
gender    False
age       False
topic     False
sign      False
date      False
text      False
dtype: bool

In [9]:
data=data.head(10000)

In [10]:
data['age']=data['age'].astype('object')

In [12]:
data['age'].nunique()

24

In [13]:
data['age'].unique()

array([15, 33, 14, 25, 17, 23, 37, 26, 24, 27, 45, 34, 41, 44, 16, 39, 35,
       36, 46, 42, 13, 38, 43, 40], dtype=object)

In [14]:
data['age'].value_counts()

age
35    2315
36    1708
17    1185
27    1054
24     655
15     602
34     553
16     440
25     386
23     253
26     234
14     212
33     136
39      79
38      46
13      42
37      33
41      20
45      16
42      14
46       7
43       6
44       3
40       1
Name: count, dtype: int64

In [15]:
data['clean_data']=data['text'].apply(lambda x: re.sub(r'[^A-Za-z]+',' ',x))

In [16]:
data['clean_data']=data['clean_data'].apply(lambda x: x.lower())

In [17]:
data['clean_data']=data['clean_data'].apply(lambda x: x.strip())

In [18]:
print("Actual data=======> {}".format(data['text'][1]))

Actual data=======>            These are the team members:   Drewes van der Laag           urlLink mail  Ruiyu Xie                     urlLink mail  Bryan Aaldering (me)          urlLink mail          


In [19]:
print("Cleaned data=======> {}".format(data['clean_data'][1]))

Cleaned data=======> these are the team members drewes van der laag urllink mail ruiyu xie urllink mail bryan aaldering me urllink mail


In [20]:
from nltk.corpus import stopwords
stopwords=set(stopwords.words('english'))

In [21]:
data['clean_data']=data['clean_data'].apply(lambda x: ' '.join([words for words in x.split() if words not in stopwords]))

In [22]:
data['clean_data'][6]

'somehow coca cola way summing things well early flagship jingle like buy world coke tune like teach world sing pretty much summed post woodstock era well add much sales catchy tune korea coke theme urllink stop thinking feel pretty much sums lot korea koreans look relaxed couple stopped thinking started feeling course high regard education math logic deep think many koreans really like work emotion anything else westerners seem sublimate moreso least display different way maybe scratch westerners koreans probably pretty similar context different anyways think losing korea repeat stop thinking feel stop thinking feel stop thinking feel everything alright'

In [23]:
data['labels']=data.apply(lambda col: [col['gender'],str(col['age']),col['topic'],col['sign']], axis=1)

In [24]:
data.head()

,id,gender,age,topic,sign,date,text,clean_data,labels
0,2059027,male,15,Student,Leo,"14,May,2004","Info has been found (+/- 100 pages,...",info found pages mb pdf files wait untill team...,"[male, 15, Student, Leo]"
1,2059027,male,15,Student,Leo,"13,May,2004",These are the team members: Drewe...,team members drewes van der laag urllink mail ...,"[male, 15, Student, Leo]"
2,2059027,male,15,Student,Leo,"12,May,2004",In het kader van kernfusie op aarde...,het kader van kernfusie op aarde maak je eigen...,"[male, 15, Student, Leo]"
3,2059027,male,15,Student,Leo,"12,May,2004",testing!!! testing!!!,testing testing,"[male, 15, Student, Leo]"
4,3581210,male,33,InvestmentBanking,Aquarius,"11,June,2004",Thanks to Yahoo!'s Toolbar I can ...,thanks yahoo toolbar capture urls popups means...,"[male, 33, InvestmentBanking, Aquarius]"


In [26]:
data=data[['clean_data','labels']]

In [27]:
data

,clean_data,labels
0,info found pages mb pdf files wait untill team...,"[male, 15, Student, Leo]"
1,team members drewes van der laag urllink mail ...,"[male, 15, Student, Leo]"
2,het kader van kernfusie op aarde maak je eigen...,"[male, 15, Student, Leo]"
3,testing testing,"[male, 15, Student, Leo]"
4,thanks yahoo toolbar capture urls popups means...,"[male, 33, InvestmentBanking, Aquarius]"
...,...,...
9995,take home forever may rest sleep arms forgotte...,"[female, 25, indUnk, Pisces]"
9996,seductive secretness behind doors warning neve...,"[female, 25, indUnk, Pisces]"
9997,kind need holding hand petting hair cry bring ...,"[female, 25, indUnk, Pisces]"
9998,blurry outside sounds people mingle pass darkn...,"[female, 25, indUnk, Pisces]"


In [28]:
X=data['clean_data']

In [29]:
Y=data['labels']

In [30]:
from sklearn.feature_extraction.text import CountVectorizer

In [31]:
vectorizer=CountVectorizer(binary=True, ngram_range=(1,2))

In [32]:
X=vectorizer.fit_transform(X)

In [33]:
X[1]

<1x643302 sparse matrix of type '<class 'numpy.int64'>'
	with 25 stored elements in Compressed Sparse Row format>

In [36]:
vectorizer.get_feature_names_out()[:5]

array(['aa', 'aa amazing', 'aa anger', 'aa compared', 'aa keeps'],
      dtype=object)

In [37]:
label_counts=dict()

for labels in data.labels.values:
    for label in labels:
        if label in label_counts:
            label_counts[label]+=1
        else:
            label_counts[label]=1

In [38]:
label_counts

{'male': 5916,
 '15': 602,
 'Student': 1137,
 'Leo': 301,
 '33': 136,
 'InvestmentBanking': 70,
 'Aquarius': 571,
 'female': 4084,
 '14': 212,
 'indUnk': 3287,
 'Aries': 4198,
 '25': 386,
 'Capricorn': 215,
 '17': 1185,
 'Gemini': 150,
 '23': 253,
 'Non-Profit': 71,
 'Cancer': 504,
 'Banking': 16,
 '37': 33,
 'Sagittarius': 1097,
 '26': 234,
 '24': 655,
 'Scorpio': 971,
 '27': 1054,
 'Education': 270,
 '45': 16,
 'Engineering': 127,
 'Libra': 491,
 'Science': 63,
 '34': 553,
 '41': 20,
 'Communications-Media': 99,
 'BusinessServices': 91,
 'Sports-Recreation': 80,
 'Virgo': 236,
 'Taurus': 812,
 'Arts': 45,
 'Pisces': 454,
 '44': 3,
 '16': 440,
 'Internet': 118,
 'Museums-Libraries': 17,
 'Accounting': 4,
 '39': 79,
 '35': 2315,
 'Technology': 2654,
 '36': 1708,
 'Law': 11,
 '46': 7,
 'Consulting': 21,
 'Automotive': 14,
 '42': 14,
 'Religion': 9,
 '13': 42,
 'Fashion': 1622,
 '38': 46,
 '43': 6,
 'Publishing': 4,
 '40': 1,
 'Marketing': 156,
 'LawEnforcement-Security': 10,
 'HumanReso

In [39]:
from sklearn.preprocessing import MultiLabelBinarizer
binarizer=MultiLabelBinarizer(classes=sorted(label_counts.keys()))

In [40]:
binarizer

MultiLabelBinarizer(classes=['13', '14', '15', '16', '17', '23', '24', '25',
                             '26', '27', '33', '34', '35', '36', '37', '38',
                             '39', '40', '41', '42', '43', '44', '45', '46',
                             'Accounting', 'Aquarius', 'Aries', 'Arts',
                             'Automotive', 'Banking', ...])

In [41]:
Y=binarizer.fit_transform(data.labels)

In [42]:
from sklearn.model_selection import train_test_split

In [43]:
Xtrain,Xtest,Ytrain,Ytest=train_test_split(X,Y,test_size=0.2)

In [44]:
from sklearn.multiclass import OneVsRestClassifier
from sklearn.linear_model import LogisticRegression

In [45]:
model=LogisticRegression(solver='lbfgs')

In [46]:
model=OneVsRestClassifier(model)

In [47]:
model.fit(Xtrain,Ytrain)

OneVsRestClassifier(estimator=LogisticRegression())

In [48]:
Ypred=model.predict(Xtest)

In [49]:
Ypred_inversed = binarizer.inverse_transform(Ypred)
y_test_inversed = binarizer.inverse_transform(Ytest)

In [50]:
for i in range(5):
    print('Text:\t{}\nTrue labels:\t{}\nPredicted labels:\t{}\n\n'.format(
        Xtest[i],
        ','.join(y_test_inversed[i]),
        ','.join(Ypred_inversed[i])
    ))

Text:	  (0, 52528)	1
  (0, 477858)	1
  (0, 567161)	1
  (0, 586020)	1
  (0, 220536)	1
  (0, 375481)	1
  (0, 465207)	1
  (0, 615971)	1
  (0, 310781)	1
  (0, 227615)	1
  (0, 621438)	1
  (0, 558082)	1
  (0, 225428)	1
  (0, 616967)	1
  (0, 251043)	1
  (0, 323620)	1
  (0, 323022)	1
  (0, 478296)	1
  (0, 323537)	1
  (0, 556303)	1
  (0, 2199)	1
  (0, 227048)	1
  (0, 342568)	1
  (0, 127378)	1
  (0, 119362)	1
  :	:
  (0, 323660)	1
  (0, 129577)	1
  (0, 406872)	1
  (0, 251149)	1
  (0, 603419)	1
  (0, 615723)	1
  (0, 419150)	1
  (0, 567774)	1
  (0, 220390)	1
  (0, 101007)	1
  (0, 357707)	1
  (0, 634892)	1
  (0, 15802)	1
  (0, 149102)	1
  (0, 227077)	1
  (0, 457172)	1
  (0, 556530)	1
  (0, 376356)	1
  (0, 541293)	1
  (0, 337228)	1
  (0, 621440)	1
  (0, 2218)	1
  (0, 128253)	1
  (0, 406847)	1
  (0, 243384)	1
True labels:	34,Sagittarius,female,indUnk
Predicted labels:	34,Sagittarius,female,indUnk


Text:	  (0, 312628)	1
  (0, 209803)	1
  (0, 607406)	1
  (0, 199958)	1
  (0, 564667)	1
  (0, 90309)	1
  

In [51]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
from sklearn.metrics import average_precision_score
from sklearn.metrics import recall_score

def print_evaluation_scores(Ytest, Ypred):
    print('Accuracy score: ', accuracy_score(Ytest, Ypred))
    print('F1 score: ', f1_score(Ytest, Ypred, average='micro'))
    print('Average precision score: ', average_precision_score(Ytest, Ypred, average='micro'))
    print('Average recall score: ', recall_score(Ytest, Ypred, average='micro'))

In [52]:
print_evaluation_scores(Ytest, Ypred)

Accuracy score:  0.317
F1 score:  0.6399517927086472
Average precision score:  0.45685002843062933
Average recall score:  0.531


*   **Low Accuracy (31.7%):** The model correctly classifies only about 32% of instances, indicating poor overall performance.
*   **Moderate F1-score (64%):** Suggests a reasonable balance between precision and recall, but still leaves room for improvement.
*   **Low Average Precision (46%):** Indicates a moderate number of false positives.
*   **Moderate Recall (53%):** The model captures slightly over half of the true positive instances.
*   **High False Positives/Negatives:** The combination of low accuracy and moderate F1 suggests a substantial number of both false positives and false negatives, with a tendency towards more false positives.
*   **Potential Data Imbalance:** The difference between precision and recall *could* indicate imbalanced data.
*   **Model Improvement Needed:** Feature engineering, trying different algorithms, hyperparameter tuning, or addressing data imbalance should be explored.


In [60]:
y_test_bin = np.array(Ytest)
y_pred_bin = np.array(Ypred)


for i in range(5):
    print("Text:", Xtest[i])
    print("True Labels:", binarizer.inverse_transform(y_test_bin[i].reshape(1, -1)))  
    print("Predicted Labels:", binarizer.inverse_transform(y_pred_bin[i].reshape(1, -1)))  
    print()

Text:   (0, 52528)	1
  (0, 477858)	1
  (0, 567161)	1
  (0, 586020)	1
  (0, 220536)	1
  (0, 375481)	1
  (0, 465207)	1
  (0, 615971)	1
  (0, 310781)	1
  (0, 227615)	1
  (0, 621438)	1
  (0, 558082)	1
  (0, 225428)	1
  (0, 616967)	1
  (0, 251043)	1
  (0, 323620)	1
  (0, 323022)	1
  (0, 478296)	1
  (0, 323537)	1
  (0, 556303)	1
  (0, 2199)	1
  (0, 227048)	1
  (0, 342568)	1
  (0, 127378)	1
  (0, 119362)	1
  :	:
  (0, 323660)	1
  (0, 129577)	1
  (0, 406872)	1
  (0, 251149)	1
  (0, 603419)	1
  (0, 615723)	1
  (0, 419150)	1
  (0, 567774)	1
  (0, 220390)	1
  (0, 101007)	1
  (0, 357707)	1
  (0, 634892)	1
  (0, 15802)	1
  (0, 149102)	1
  (0, 227077)	1
  (0, 457172)	1
  (0, 556530)	1
  (0, 376356)	1
  (0, 541293)	1
  (0, 337228)	1
  (0, 621440)	1
  (0, 2218)	1
  (0, 128253)	1
  (0, 406847)	1
  (0, 243384)	1
True Labels: [('34', 'Sagittarius', 'female', 'indUnk')]
Predicted Labels: [('34', 'Sagittarius', 'female', 'indUnk')]

Text:   (0, 312628)	1
  (0, 209803)	1
  (0, 607406)	1
  (0, 199958)	1
  (0